File: Daily_Lists/search_in_a_Big_sorted_array/search_in_a_big_sorted_array.java
/**
 * Definition of ArrayReader:
 *
 * class ArrayReader {
 *      // get the number at index, return -1 if index is less than zero.
 *      public int get(int index);
 * }
 */
public class Solution {
    /**
     * @param reader: An instance of ArrayReader.
     * @param target: An integer
     * @return : An integer which is the index of the target number
     * @link: http://www.lintcode.com/en/problem/search-in-a-big-sorted-array/
     * @author: Egbert Li
     */
    public int searchBigSortedArray(ArrayReader reader, int target) {
        int index = 1;
        while (target > reader.get(index - 1)) {
            index *= 2;
        }
        int start = 0;
        int end = index - 1;
        while (start + 1 < end) {
            int mid = start + (end -start) / 2;
            if (target > reader.get(mid)) {
                start = mid;
            } else {
                end = mid;
            }
        }
        if (target == reader.get(start)) {
            return start;
        } else if (target == reader.get(end)) {
            return end;
        } else {
            return -1;
        }
    }
}

File: Boxuan/16_01_17/Search_in_a_Big_Sorted_Array.py
class Solution:
    # @param {int[]} A an integer array
    # @param {int} target an integer
    # @return {int} an integer
    def searchBigSortedArray(self, A, target):
        if not A:
            return -1
        right = 0

        while right < len(A) and A[right] < target:
            right *= 2
            right += 1
        if right >= len(A):
            right = len(A) - 1
        
        left = 0
        while left + 1 < right:
            mid = left + (right - left) / 2
            if A[mid] < target:
                left = mid
            else:
                right = mid

        if A[left] == target:
            return left
        if A[right] == target:
            return right
            
        return -1

        